Self sign-up for paid plans: show banner when pending selection, complete workflow when plan selected

nospame opened the pull request:
## Product Description
When a self-signup workflow is active, this banner will appear on the dashboard for users who have edit billing permission.
<img width="1155" alt="image" src="https://github.com/user-attachments/assets/4bdc2f2f-eac4-4c23-88fb-7ef4187fa681">


## Technical Summary
[SAAS-16030](https://dimagi.atlassian.net/browse/SAAS-16030), [SAAS-16031](https://dimagi.atlassian.net/browse/SAAS-16031)
Adds a banner template styled to be visually similar to the paused plan notice, as well as a `get_pending_plan_context` function called in the same way/location as `get_paused_plan_context`.

Adds functionality for an in-progress `SelfSignupWorkflow` to be completed in two places:
- from this banner, using a small form via the new `dismiss_self_signup` view (will always select Community because a user is essentially opting out of the workflow)
- when a paid plan is selected and confirmed through the regular _select plan -> confirm plan -> billing info_ steps

Also memoized `get_in_progress_for_domain` so we don't make a db call every time the dashboard view is loaded, and renamed some "paused" css styles to be more generic.

None of this is accessible in practice yet, until changes to the initial account registration views are made that will initiate the `SelfSignupWorkflow`.

## Feature Flag

## Safety Assurance

### Safety story
 Tested locally to see that the banner loads correctly with appropriate context based on whether the `SelfSignupWorkflow` is active and whether the user has permissions to actually select a plan for the domain. Also confirmed that the `SelfSignupWorkflow` is closed out correctly based on the the "Don't show" option in the banner, and when a paid plan is subscribed to.

### Automated test coverage
I haven't added automated testing here, but the memoization of `get_in_progress_for_domain` needed a small change to previously added tests.

### QA Plan
This will get QA as part of the Self-signup for paid software plans feature.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [ ] Risk label is set correctly
  - I believe the auto-added "Risk: High" is misleading here and likely only because a file in the `registration` app was changed. The actual risk is likely much lower.
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SAAS-16030]: https://dimagi.atlassian.net/browse/SAAS-16030?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SAAS-16031]: https://dimagi.atlassian.net/browse/SAAS-16031?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nospame committed:
Add pending plan signup notice

nospame committed:
Rename paused-plan stylesheets

nospame committed:
Add dismiss_self_signup to pending plan notice

nospame committed:
Memoize get_in_progress_for_domain

nospame committed:
"Bootstrap 5 Migration - Rebuilt diffs"

nospame committed:
Update test to clear SelfSignupWorkflow

nospame committed:
Complete self-signup workflow when subscription is confirmed

orangejenny commented on corehq/apps/dashboard/templates/dashboard/base.html:
```diff
@@ -25,6 +25,8 @@ <h1 class="mb-3 mt-5">
     {% if is_paused %}
       {% include 'hqwebapp/partials/bootstrap5/paused_plan_notice.html' %}
+    {% elif should_show_pending_notice %}
+      {% include 'hqwebapp/partials/bootstrap5/pending_plan_notice.html' %}
```
Can you move `pending_plan_notice.html` up a directory? Only files that have both a B3 and B5 version should be in bootstrap3/bootstrap5 directories.

orangejenny commented on corehq/apps/dashboard/views.py:
```diff
@@ -273,3 +281,14 @@ def apps_link(urlname, req):
+def dismiss_self_signup(request, domain):
+    self_signup = SelfSignupWorkflow.get_in_progress_for_domain(domain)
+    if self_signup is not None:
+        self_signup.complete_workflow(SoftwarePlanEdition.COMMUNITY)
```
Could the user get feedback that they're now "subscribed" to community? It could be something simple like using `messages`.

nospame replied:
Yep, totally reasonable. Made it "Subscribed to Community Edition." but can easily change if needed (expecting there may be feedback on specific wordings from Ops et al. during QA).

orangejenny commented on corehq/apps/registration/models.py:
```diff
@@ -170,6 +170,7 @@ class SelfSignupWorkflow(models.Model):
     )
 
     @classmethod
+    @memoized
```
I think you want quickcache for this. Memoized only lasts for the duration of the request. You could probably quickcache this value for quite a while.

orangejenny commented on corehq/apps/registration/models.py:
```diff
@@ -184,3 +185,4 @@ def complete_workflow(self, edition):
         self.completed_on = datetime.datetime.now()
         self.subscribed_edition = edition
         self.save()
+        self.get_in_progress_for_domain.reset_cache(self.domain)
```
If `get_in_progress_for_domain` is cached (regardless of memoization vs quickcache), shouldn't that also be cleared when a new `SelfSignupWorkflow` is created?

nospame replied:
Not in any current or planned implementation, since the workflow object is only created at the same time the domain is, here (part of `request_new_domain`): https://github.com/dimagi/commcare-hq/blob/41e0db5b51c5c3582e2d6b522de115b569e62ac3/corehq/apps/registration/utils.py#L169-L170

orangejenny reviewed (commented):
This looks good, nice attention to detail. Left a couple of comments.

mjriley commented on corehq/apps/hqwebapp/templates/hqwebapp/partials/bootstrap5/pending_plan_notice.html:
```diff
@@ -23,12 +23,15 @@ <h3>
+      <form action="{% url 'dismiss_self_signup' domain %}" method="POST">
+        {% csrf_token %}
+        <p class="small">
+          {% blocktrans %}
+            Prefer to stay with Community Edition instead?
+            <button class="btn btn-sm btn-link p-0" type="submit">Don't show this banner again.</button>
+          {% endblocktrans %}
```
This can probably be separated into two separate translation tags

jingcheng16 commented:
Reviewed with @mjriley in product code review session. Overall it looks good. I can go back to it once all feedbacks is addressed.

nospame committed:
PR feedback:
- Move pending_plan_notice out of bootstrap5 dir
- Add message when user dismisses banner
- Use quickcache not memoized
- Use trans not blocktrans
